README.md
# shellcgi - Shell framework for CGI programming

Copyright (C) 2009 Masahiko Ito

These programs is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or (at
your option) any later version.

These programs is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License along
with these programs; if not, write to the Free Software Foundation, Inc.,
59 Temple Place, Suite 330, Boston, MA 02111-1307 USA

Mail suggestions and bug reports for these programs to
"Masahiko Ito" \<m-ito@myh.no-ip.org\>

## 目的
一般的にCGIプログラムを作成する際には、ruby,perl,python,php等
々のスクリプト言語を 利用することが多いかと思いますが、shellcgi
はshell(bash)スクリプトでCGIプログラムを 構築することを容易に
するフレームワークです。shellcgiはメインフレームコンピュータ
(Z/OS、VOS3、ACOS-4 etc)での会話処理プログラムの構造に近い形
でCGIプログラムを実現します。 たとえば、1本のプログラムは1つ
の画面の作成及び処理を行います。画面遷移(プログラム遷移)はト
ランザクションコードで指定し、トランザクションコードに紐づけ
られたプログラムが順次 実行されます。各プログラム間で受け渡す
必要のあるデータは、SPA(Scratch Pad Area)を介して 連携します。
このような構造でプログラミングが可能なことで、従来COBOL等でシ
ステム開発 をしてきたエンジニアでもCGIを利用したWebアプリケー
ションを作成しやすくなります。

## インストール方法(例)
```
tar xvf shellcgi.tar.gz
cp -r shellcgi/public_html/shellcgi ~/public_html/
cp -r shellcgi/.shellcgi ~/
```

## 環境設定
~/public_html/shellcgi/cgi-bin/shellcgi.sh内の環境変数を適切に設定します。
最低限、**CGICTRL_HOMEDIR** の **_USERNAME_** 部分は自己環境に合わせて修正が必要です。
```
CGICTRL_HOMEDIR=/home/_USERNAME_/.shellcgi; export CGICTRL_HOMEDIR
CGICTRL_TMPDIR=/tmp/.shellcgi; export CGICTRL_TMPDIR
CGICTRL_LOGGING=YES; export CGICTRL_LOGGING
CGICTRL_NOAUTHENTICATEDUSER=anonymous; export CGICTRL_NOAUTHENTICATEDUSER
CGICTRL_MAXLOCKRETRY=180; export CGICTRL_MAXLOCKRETRY
CGICTRL_SWEEPDAY=2; export CGICTRL_SWEEPDAY
```

## 開発の手順(例)

### 実装する機能
今回はサンプルとして、

* 文字列を入力させる
* /etc/shellsから入力させた文字列を含む行を検索する
* 検索された行を表示する
* 空文字列が入力された場合はエラーメッセージを表示する

というのを基本的な機能とします。画面構成は文字列を入力させる画面と、
検索結果を表示する画面の2画面構成とします。

### 画面を設計する
~/.shellcgi/html/の下にhtmlファイル形式で画面を作成します。

#### 入力画面(SAMPLE1.html)
```
 1: Content-Type: text/html
 2: 
 3: <html>
 4: <head>
 5:     <meta http-equiv="Content-Type" content="text/html;charset=UTF-8">
 6:     <title>検索文字列を入力してください</title>
 7: </head>
 8: <body>
 9:     [sample1]
10:     <h1>検索文字列を入力してください</h1>
11: 
12:     <form action="./shellcgi.cgi" method="POST" enctype="multipart/form-data">
13:     <input type="hidden" value="@{CGICTRL_SESSIONID}@" name="CGICTRL_SESSIONID">
14:     <input type="text" size="80" value="@{search}@" maxlength="80" name="search"><br>
15:     <input type="submit" value="検索開始" name="submit">
16:     </form>
17:     <pre style="color:@{msgcolor=black}@">
18: @{message=}@
19:     </pre>
20: </body>
21: </html>
```
行番号は説明の便宜上付番しているものなので、実際に記述してはいけません。
では上から順に説明します。
* 1: Content-typeを指定します。通常はサンプルの通りで良いでしょう。
* 2: 必ず改行させます。
* 3: この行以降に基本的にはhtml文法に従って定義していきます。
* 12: formタグのactionには必ず **shellcgi.cgi** を呼び出すように指定します。
* 13: formタグには必ずこのhiddenタイプのinputタグを記述します。通常はサンプルの通りで良いでしょう。
* 14: **@{search}@** 部分はCGIプログラムから置き換えます。
* 17: **@{msgcolor=black}@** 部分はCGIプログラムから置き換えます。CGIプログラムから置き換えなかった場合は **black** が規定値となります。
* 18: **@{message=}@** 部分はCGIプログラムから置き換えます。規定値として空文字が指定されています。

#### 検索結果画面(SAMPLE2.html)
```
 1: #
 2: # START=@:
 3: # END=:@
 4: #
 5: Content-Type: text/html
 6: 
 7: <html>
 8: <head>
 9:     <meta http-equiv="Content-Type" content="text/html;charset=UTF-8">
10:     <title>検索結果</title>
11: </head>
12: <body>
13:     [sample2]
14:     <h1>検索結果</h1>
15: 
16:     <form action="./shellcgi.cgi" method="POST" enctype="multipart/form-data">
17:     <input type="hidden" value="@:CGICTRL_SESSIONID:@" name="CGICTRL_SESSIONID">
18:     <input type="submit" value="戻る" name="submit">
19:     </form>
20:     <h2>検索文字列</h2>
21:     <pre>
22: @:search:@
23:     </pre>
24:     <h2>検索結果</h2>
25:     <pre>
26: @:result:@
27:     </pre>
28:     <pre>
29: @:message=:@
30:     </pre>
31: </body>
32: </html>
```
行番号は説明の便宜上付番しているものなので、実際に記述してはいけません。
では上から順に説明します。
* 1: 「#」で始まる行はコメント行として扱われます。
* 2: コメント行に、 **START=** を記述した場合は、 **=** の右辺値で置換対象文字列の開始を表す **@{** を変更します。
* 3: コメント行に、 **END=** を記述した場合は、 **=** の右辺値で置換対象文字列の終了を表す **}@** を変更します。
* 5: Content-typeを指定します。通常はサンプルの通りで良いでしょう。
* 6: 必ず改行させます。
* 7: この行以降に基本的にはhtml文法に従って定義していきます。
* 16: formタグのactionには必ず **shellcgi.cgi** を呼び出すように指定します。
* 17: formタグには必ずこのhiddenタイプのinputタグを記述します。通常はサンプルの通りで良いでしょう。
* 14: **@:search:@** 部分はCGIプログラムから置き換えます。
* 26: **@:result:@** 部分はCGIプログラムから置き換えます。
* 29: **@:message=:@** 部分はCGIプログラムから置き換えます。規定値として空文字が指定されています。

### リソースを定義する
CGIプログラムで排他制御の対象とするリソースを定義します。
~/.shellcgi/resource/の下にリソースを示す文字列(リソースID)と同名のファイルを作成し、その中に
リソースの実体を示す文字列(ファイルであればフルパス等)を記述します。
```
echo "/etc/shells" >~/.shellcgi/resource/ETC_SHELLS
```

### トランザクションとCGIプログラムを定義する
~/.shellcgi/tranpgm/の下にトランザクションを示す文字列(トランザクションID)と同名のファイルを作成し、その中に
トランザクションに紐づくCGIプログラムを記述します。
```
echo "${HOME}/public_html/shellcgi/cgi-bin/sample1.cgi" >~/.shellcgi/tranpgm/SAMPLE1
echo "${HOME}/public_html/shellcgi/cgi-bin/sample2.cgi" >~/.shellcgi/tranpgm/SAMPLE2
```

### トランザクションとリソースを定義する
\~/.shellcgi/tranres/の下にトランザクションを示す文字列(トランザクションID)と同名のファイルを作成し、その中に
トランザクションで排他的に使用するリソースのリソースIDを記述します。リソースIDは\~/.shellcgi/resource/の下に定義
されている必要があります。
```
echo "ETC_SHELLS" >~/.shellcgi/tranres/SAMPLE1
echo "ETC_SHELLS" >~/.shellcgi/tranres/SAMPLE2
```

### ユーザに許可するトランザクションを定義する
~/.shellcgi/usertran/の下にBASIC認証またはDIGEST認証するユーザ名と同名のファイルを作成し、その中に
そのユーザに許可するトランザクションIDを記述します(正規表現可能)。
認証機構を使用しない場合は全てのユーザが仮想的にユーザ **anonymous** でアクセスしていることになるので、
**anonymous** に許可するトランザクションIDを記述する必要があります。
```
echo "^SAMPLE[01]$" >~/.shellcgi/usertran/anonymous
```

### sample1.cgi(入力画面処理CGI)
```
 1: #! /bin/sh
 2: . ./shellcgi.sh
 3: #-- USER CODING START(1)
 4: #-- USER CODING END(1)
 5: if [ `isinputmode` ]
 6: then
 7: #
 8: # This block processes input data from FORM or SPA and determine next transaction which create screen.
 9: #
10: #-- USER CODING START(2)
11:     search=`getvalue search`
12:     if [ "X${search}" = "X" ]
13:     then
14:         setspa search "${search}"
15:         setspa message "検索文字列が指定されていません"
16:         setspa msgcolor "red"
17:         settran "SAMPLE1"
18:     else
19:         user=`getuser`
20:         if [ `ispermittedwith "${user}" SAMPLE2` ]
21:         then
22:             setspa search "${search}"
23:             setspa message ""
24:             setspa msgcolor "black"
25:             settran "SAMPLE2"
26:         else
27:             setspa search "${search}"
28:             setspa message "${user}は結果の表示を許可されていません"
29:             setspa msgcolor "red"
30:             settran "SAMPLE1"
31:         fi
32:     fi
33: #-- USER CODING END(2)
34: else
35: #
36: # This block edit screen from SPA and output it.
37: # Next transaction must be a transaction for this program.
38: #
39: #-- USER CODING START(3)
40:     thistran=`gettran`
41:     search=`getspa search`
42:     message=`getspa message`
43:     msgcolor=`getspa msgcolor`
44:     outhtml ${thistran}.html \
45:         "search=${search}" \
46:         "message=${message}" \
47:         "msgcolor=${msgcolor}"
48:     settran "${thistran}"
49: #-- USER CODING END(3)
50: fi
51: exit 0
```
* 原則、**#-- USER CODING START(x)〜#-- USER CODING END(x)** の間にコーディングを行います。
* **USER CODING START(1)** のブロック内は任意のコーディングを行います。
* **USER CODING START(2)** のブロック内は **form** から入力されたデータを処理し、
次に展開するトランザクションを決定( **settran** )し、そのトランザクションに引き継ぐデータをSPA(Scratch Pad Area)に格納( **setspa** )します。
* **USER CODING START(3)** のブロック内は **SPA** 等からのデータを利用して画面を編集し出力( **outhtml** )します。

### sample2.cgi(検索結果画面処理CGI)
```
 1: #! /bin/sh
 2: . ./shellcgi.sh
 3: #-- USER CODING START(1)
 4: #-- USER CODING END(1)
 5: if [ `isinputmode` ]
 6: then
 7: #
 8: # This block processes input data from FORM or SPA and determine next transaction which create screen.
 9: #
10: #-- USER CODING START(2)
11:     settran "SAMPLE1"
12: #-- USER CODING END(2)
13: else
14: #
15: # This block edit screen from SPA and output it.
16: # Next transaction must be a transaction for this program.
17: #
18: #-- USER CODING START(3)
19:     thistran=`gettran`
20:     search=`getspa search`
21:     inputfile=`getrealres "ETC_SHELLS"`
22:     result=`cat ${inputfile} | egrep "${search}"`
23:     setspa search "${search}"
24:     outhtml ${thistran}.html \
25:         "search=${search}" \
26:         "result=${result}"
27:         "message=" \
28:     settran "${thistran}"
29: #-- USER CODING END(3)
30: fi
31: exit 0
```

### ファンクション
* getvalue 項目名  
  **form** タグからの入力項目を取得する  
  **項目名** は **input** タグ等の **name** パラメータ値
* getfilename 項目名  
  \<input type="file" ... \> のファイル名を取得する
* outhtml HTMLファイル名 項目名1=値1 項目名2=値2 ...  
  HTMLファイルの **@{項目名n}@** を **値n** に置き換えて出力する  
* getrealres リソースID  
  **リソースID** に設定された実体を示す文字列を取得する
* getspa 項目名  
  SPAに設定した値を取得する
* setspa 項目名 設定値  
  SPAに値を設定する
* clearallspa  
  自セッション内のSPAを全て削除する
* clearspa 項目名  
  自セッション内の **項目名** で指定したSPAを削除する
* isoutputmode  
  画面(html)を出力するべきモードの場合 **空文字列** 以外を取得し、$?=0を返す
* isinputmode  
  画面(html)から受信するべきモードの場合 **空文字列** 以外を取得し、$?=0を返す
* getuser  
  ユーザ名を取得する  
  非認証モードの場合 **anonymous** を取得する
* settran トランザクション名  
  次に実行するべきトランザクションを指定する
* gettran  
  現在のトランザクション名を取得する
* ispermitted  
  現在のユーザ、トランザクションの組み合わせが許可されている場合、**空文字列** 以外を取得し、$?=0を返す
* ispermittedwith ユーザ名 トランザクション名  
  指定したユーザ、トランザクションの組み合わせが許可されている場合、**空文字列** 以外を取得し、$?=0を返す

### 起動の仕方
会話処理の一番最初はGETメソッドを使い、パラメータ「inittran」に
初期トランザクションを指定して、「shellcgi.cgi」を呼び出します。
```
 1: <html>
 2: <head>
 3:     <META HTTP-EQUIV="Content-Type" CONTENT="text/html;charset=UTF-8">
 4:     <title>shellcgi sample page</title>
 5: </head>
 6: <body>
 7:     <h1>shellcgi sample page</h1>
 8:     <ul>
 9:         <li><a href="./cgi-bin/shellcgi.cgi?inittran=SAMPLE1">shellcgi sample</a>
10:     </ul>
11: </body>
12: </html>
```

## デモ
[shellcgi sample page](http://myh.no-ip.org/~m-ito/shellcgi/)


CHANGES
2022/05/07 Initial release.



public_html/shellcgi/cgi-bin/shellcgi.sh
#
# shellcgi - Shell framework for CGI programming
# Copyright (C) 2022 "Masahiko Ito"<m-ito@myh.no-ip.org>
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
#------------------------------------------------------------
# Support script for shellcgi system
#------------------------------------------------------------
#
# Set your home directory for shellcgi system.
# This directory must be readable from httpd!
#
#  ${CGICTRL_HOMEDIR}/html/_screen_name_.html
#  ${CGICTRL_HOMEDIR}/resource/_resourcename_
#  ${CGICTRL_HOMEDIR}/tranpgm/_tranname_
#  ${CGICTRL_HOMEDIR}/tranres/_tranname_
#  ${CGICTRL_HOMEDIR}/usertran/_username_
#
CGICTRL_HOMEDIR=/home/_USERNAME_/.shellcgi; export CGICTRL_HOMEDIR
#
# Set temporary directory for shellcgi system.
# This directory will be created by shellcgi.cgi automatically.
# You should not change CGICTRL_TMPDIR without understanding what you are doing!
#
CGICTRL_TMPDIR=/tmp/.shellcgi; export CGICTRL_TMPDIR
#
# Set YES to get access log or set NO to be careless about access log
#
CGICTRL_LOGGING=YES; export CGICTRL_LOGGING
#
# Set user name who is not authenticated
#
CGICTRL_NOAUTHENTICATEDUSER=anonymous; export CGICTRL_NOAUTHENTICATEDUSER
#
# Set maximum count for lock retry(1 retry is 1 second)
#
CGICTRL_MAXLOCKRETRY=180; export CGICTRL_MAXLOCKRETRY
#
# Set sweepday. Old spa and input which was created sweepday before will be sweeped.
#
CGICTRL_SWEEPDAY=2; export CGICTRL_SWEEPDAY
#------------------------------------------------------------
# Functions for shellcgi system
#------------------------------------------------------------
#
# Usage: startcgi
# Initialize environment for starting cgi procedure. 
# This gets input data into temporary file from environment value `QUERY_STRING' or stdin.
#
function startcgi {
	umask 077
	if [ ! -d "${CGICTRL_TMPDIR}" ]
	then
		mkdir "${CGICTRL_TMPDIR}"
		mkdir "${CGICTRL_TMPDIR}/spa"
		mkdir "${CGICTRL_TMPDIR}/input"
		mkdir "${CGICTRL_TMPDIR}/param"
		mkdir "${CGICTRL_TMPDIR}/log"
		mkdir "${CGICTRL_TMPDIR}/lock"
	fi
	CGICTRL_UNIQUEID="${REMOTE_ADDR}".`date +"%Y%m%d%H%M%S.%N"`."${RANDOM}".$$; export CGICTRL_UNIQUEID
	CGICTRL_INPUT="${CGICTRL_TMPDIR}/input/${CGICTRL_UNIQUEID}"; export CGICTRL_INPUT
	if [ "X${REQUEST_METHOD}" = "XGET" ]
	then
		echo "${QUERY_STRING}" >"${CGICTRL_INPUT}"
	else
		dd bs="${CONTENT_LENGTH}" count=1 >"${CGICTRL_INPUT}" 2>/dev/null
		CGICTRL_SEP=`cat "${CGICTRL_INPUT}" | head -1`; export CGICTRL_SEP
	fi
}
#
# Usage: endcgi
# Clear environment for cgi procedure. 
# This remove temporary files.
#
function endcgi {
	rm "${CGICTRL_INPUT}"
}
#
# Usage: newsessionid
# Get new sessionid string
#
function newsessionid {
	mkdir "${CGICTRL_TMPDIR}/spa/${CGICTRL_UNIQUEID}"
	echo "${CGICTRL_UNIQUEID}"
}
#
# Usage: getpgm transaction
# Get fullpath of cgi program related to transaction
#
function getpgm {
	tran="$1"
	pgm=`cat "${CGICTRL_HOMEDIR}/tranpgm/${tran}" | egrep -v '^#' | sed -e '/^$/d'`
	echo "${pgm}"
}
#
# Usage: sweepspa days
# Clear spa cleated at days before
#
function sweepspa {
	(
		day="$1"
		cd "${CGICTRL_TMPDIR}/spa/" || return
		CGICTRL_DELDAYTIME=`date --date="${day} day ago" +"%Y%m%d999999"`; export CGICTRL_DELDAYTIME
		/bin/ls -1 |\
		egrep '^[0-9][0-9]*\.[0-9][0-9]*\.[0-9][0-9]*\.[0-9][0-9]*\.[0-9][0-9]*\.[0-9][0-9]*\.[0-9][0-9]*\.[0-9][0-9]*$'|\
		awk 'BEGIN{
			FS="."
		}
		{
			if ($5 <= ENVIRON["CGICTRL_DELDAYTIME"]){
				print
			}
		}
		END{
		}' |\
		xargs rm -fr
	)
}
#
# Usage: sweepinput days
# Clear input cleated at days before
#
function sweepinput {
	(
		day="$1"
		cd "${CGICTRL_TMPDIR}/input/" || return
		CGICTRL_DELDAYTIME=`date --date="${day} day ago" +"%Y%m%d999999"`; export CGICTRL_DELDAYTIME
		/bin/ls -1 |\
		egrep '^[0-9][0-9]*\.[0-9][0-9]*\.[0-9][0-9]*\.[0-9][0-9]*\.[0-9][0-9]*\.[0-9][0-9]*\.[0-9][0-9]*\.[0-9][0-9]*$'|\
		awk 'BEGIN{
			FS="."
		}
		{
			if ($5 <= ENVIRON["CGICTRL_DELDAYTIME"]){
				print
			}
		}
		END{
		}' |\
		xargs rm -fr
	)
}
#
# Usage: getlogdir
# Get directory name for logging
#
function getlogdir {
	echo "${CGICTRL_TMPDIR}/log"
}
#
# Usage: isloggingmode
# Get current logging mode
#
function isloggingmode {
	if [ "X${CGICTRL_LOGGING}" = "XYES" ]
	then
		echo "${CGICTRL_LOGGING}"
		return 0
	else
		return 1
	fi
}
#
# Usage: lock transaction resource
# Make lock directory for resource
#
function lock {
	tran="$1"
	res="$2"
	if [ -f ${CGICTRL_HOMEDIR}/resource/${res} ]
	then
		mkdir "${CGICTRL_TMPDIR}/lock/${res}" 2>/dev/null || return 1
		echo "${CGICTRL_UNIQUEID}" >"${CGICTRL_TMPDIR}/lock/${res}/${tran}"
		return 0
	fi
	return 2
}
#
# Usage: unlock transaction resource
# Remove lock directory for resource
#
function unlock {
	tran="$1"
	res="$2"
	if [ -f ${CGICTRL_HOMEDIR}/resource/${res} ]
	then
		ssid=`cat "${CGICTRL_TMPDIR}/lock/${res}/${tran}" 2>/dev/null`
		if [ "X${ssid}" = "X${CGICTRL_UNIQUEID}" ]
		then
			rm -fr "${CGICTRL_TMPDIR}/lock/${res}" 2>/dev/null
			return 0
		fi
		return 1
	fi
	return 2
}
#
# Usage: lockall transaction
# Make lock directory for all resources which transaction use
#
function lockall {
	tran="$1"
	for i in `cat "${CGICTRL_HOMEDIR}/tranres/${tran}" 2>/dev/null | egrep -v '^#' | sed -e '/^$/d' | sort`
	do
		count=0
		lock ${tran} "$i"
		lockstat=$?
		count=`expr ${count} + 1`
		while [ ${lockstat} -ne 0 ]
		do
			if [ ${lockstat} -eq 2 ]
			then
				unlockall ${tran}
				return 2
			fi
			if [ ${count} -ge ${CGICTRL_MAXLOCKRETRY} ]
			then
				unlockall ${tran}
				return 1
			fi
			sleep 1
			lock ${tran} "$i"
			lockstat=$?
			count=`expr ${count} + 1`
		done
	done
	return 0
}
#
# Usage: unlockall transaction
# Remove lock directory for all resources which transaction use
#
function unlockall {
	tran="$1"
	ret=0
	for i in `cat "${CGICTRL_HOMEDIR}/tranres/${tran}" 2>/dev/null | egrep -v '^#' | sed -e '/^$/d' | sort`
	do
		unlock ${tran} "$i"
		if [ $? -ne 0 ]
		then
			ret=1
		fi
	done
	return "${ret}"
}
#
# Usage: showtimeover
# Show lock failure message in html
#
function showtimeover {
	echo "Content-Type: text/html"
	echo ""
	echo "Time over while executing ${CGICTRL_TRAN}."
}
#
# Usage: showresourceerror
# Show lock failure message in html
#
function showresourceerror {
	echo "Content-Type: text/html"
	echo ""
	echo "ResourceID error while excuting ${CGICTRL_TRAN}."
}
#------------------------------------------------------------
# Functions for user applications
#------------------------------------------------------------
#
# Usage: getvalue name
# Get value from input item on <form> and output it stdout.
# <input type="text">, <input type="file">, <textarea> etc are acceptable.
#
function getvalue {
	name="$1"
	CGICTRL_CR=`echo -e -n '\r'`; export CGICTRL_CR
	if [ "X`echo ${CONTENT_TYPE} | egrep 'multipart/form-data'`" = "X" ]
	then
		for i in `cat "${CGICTRL_INPUT}" | tr '&' ' '`
		do
			inname=`echo "$i" | cut -d= -f1`
			if [ "X${inname}" = "X${name}" ]
			then
				raw_value=`echo "$i" | cut -d= -f2 | sed -e 's/+/ /g;s/%/\\\\x/g'`
				printf "${raw_value}"
			fi
		done
	else
		start=0
		end=0
		for i in `cat "${CGICTRL_INPUT}" | egrep -a -n -e "${CGICTRL_SEP}" | cut -d: -f1`
		do
			if [ ${start} -eq 0 ]
			then
				start=`expr $i + 1`
			elif [ ${end} -eq 0 ]
			then
				end=`expr $i - 1`
			else
				start=`expr ${end} + 2`
				end=`expr $i - 1`
			fi
		
			if [ ${start} -ne 0 -a ${end} -ne 0 ]
			then
				res=`sed -n -e "${start},${end}p" "${CGICTRL_INPUT}" |\
					awk '{
						if ($0 ~ /^\r$/){
							exit;
						}else{
							print;
						}
					}' |\
					egrep "Content-Disposition:.* name=\"${name}\""`
				if [ "X${res}" != "X" ] 
				then
					bias=`sed -n -e "${start},${end}p" "${CGICTRL_INPUT}" |\
						awk 'BEGIN{
							count = 0;
						}
						{
							if ($0 ~ /^\r$/){
								count++;
								print count;
								exit;
							}else{
								count++;
							}
						}'`
					start2=`expr ${start} + ${bias}`
					(
						if [ ${start2} -eq ${end} ]
						then
							sed -n -e "${start2},${end}p" "${CGICTRL_INPUT}" |\
							sed -e "s/${CGICTRL_CR}$//" |\
							tr -d '\n'
						else
							end2=`expr ${end} - 1`
							sed -n -e "${start2},${end2}p" "${CGICTRL_INPUT}"
							sed -n -e "${end},${end}p" "${CGICTRL_INPUT}" |\
							sed -e "s/${CGICTRL_CR}$//" |\
							tr -d '\n'
						fi
					)
				fi
			fi
		done
	fi
}
#
# Usage: getfilename name
# Get filename from <input type="file"> on <form> and output it stdout.
#
function getfilename {
	name="$1"
	if [ "X`echo ${CONTENT_TYPE} | egrep 'multipart/form-data'`" = "X" ]
	then
		for i in `cat "${CGICTRL_INPUT}" | tr '&' ' '`
		do
			inname=`echo $i | cut -d= -f1`
			if [ "X${inname}" = "X${name}" ]
			then
				raw_value=`echo "$i" | cut -d= -f2 | sed -e 's/+/ /g;s/%/\\\\x/g'`
				printf "${raw_value}"
			fi
		done
	else
		start=0
		end=0
		for i in `cat "${CGICTRL_INPUT}" | egrep -a -n -e "${CGICTRL_SEP}" | cut -d: -f1`
		do
			if [ ${start} -eq 0 ]
			then
				start=`expr $i + 1`
			elif [ ${end} -eq 0 ]
			then
				end=`expr $i - 1`
			else
				start=`expr ${end} + 2`
				end=`expr $i - 1`
			fi
			if [ ${start} -ne 0 -a ${end} -ne 0 ]
			then
				res=`sed -n -e "${start},${end}p" "${CGICTRL_INPUT}" |\
					awk '{
						if ($0 ~ /^\r$/){
							exit;
						}else{
							print;
						}
					}' |\
					egrep "Content-Disposition:.* name=\"${name}\""`
				if [ "X${res}" != "X" ] 
				then
					echo "${res}" | sed -e 's/^.*filename="//;s/".*$//'
				fi
			fi
		done
	fi
}
#
# Usage: outhtml htmlfile name1=value1 name2=value2 ...
# Replace name to value in htmlfile and output it stdout.
#
# Only first htmlfile in cgi must be startted with next 2 lines.
# Content-Type: text/html<NewLine>
# <NewLine>
#
function outhtml {
	CGICTRL_PARAM="${CGICTRL_TMPDIR}/param/${CGICTRL_UNIQUEID}"; export CGICTRL_PARAM
	CGICTRL_PARAMSEP="${RANDOM}${RANDOM}${RANDOM}${RANDOM}"; export CGICTRL_PARAMSEP
	html="$1"
	echo "${CGICTRL_PARAMSEP}" >"${CGICTRL_PARAM}"
	echo "CGICTRL_SESSIONID" >>"${CGICTRL_PARAM}"
	echo "${CGICTRL_SESSIONID}" >>"${CGICTRL_PARAM}"
	outsw="off"
	while [ "$#" != "0" ]
	do
		if [ "X${outsw}" = "Xon" ]
		then
			echo "${CGICTRL_PARAMSEP}" >>"${CGICTRL_PARAM}"
			echo "$1" | head -1 | sed -e 's/=.*$//' >>"${CGICTRL_PARAM}"
			echo "$1" | head -1 | sed -e 's/^[^=]*=//' >>"${CGICTRL_PARAM}"
			echo "$1" | sed -n -e '2,$p' >>"${CGICTRL_PARAM}"
		else
			outsw="on"
		fi
        	shift
	done
	CGICTRL_START_REPLACE=`cat "${CGICTRL_HOMEDIR}/html/${html}" | expand | egrep '^# *START=' | cut -d'=' -f2 | cut -d' ' -f1`; export CGICTRL_START_REPLACE
	CGICTRL_END_REPLACE=`cat "${CGICTRL_HOMEDIR}/html/${html}" | expand | egrep '^# *END=' | cut -d'=' -f2 | cut -d' ' -f1`; export CGICTRL_END_REPLACE
	cat "${CGICTRL_HOMEDIR}/html/${html}" |\
	egrep -v '^#' |\
	awk 'BEGIN{
		start = "@{"
		if (ENVIRON["CGICTRL_START_REPLACE"] != ""){
			start = ENVIRON["CGICTRL_START_REPLACE"]
		}
		end = "}@"
		if (ENVIRON["CGICTRL_END_REPLACE"] != ""){
			end = ENVIRON["CGICTRL_END_REPLACE"]
		}
		count = -1
		ret = getline <ENVIRON["CGICTRL_PARAM"]
		while (ret == 1){
			if ($0 == ENVIRON["CGICTRL_PARAMSEP"]){
				ret = getline <ENVIRON["CGICTRL_PARAM"]
				count++;
				str1a1[count] = sprintf("%s%s%s", start,$0,end)
				str1a2[count] = sprintf("%s%s=[^%s]*%s", start,$0,end,end)
				str2a[count] = ""
			}else{
				gsub(/&/,"\\\\\\&amp;")
				gsub(/</,"\\\\\\&lt;")
				gsub(/>/,"\\\\\\&gt;")
				gsub(/\x27/,"\\\\\\&#39;")
				gsub(/"/,"\\\\\\&quot;")
				gsub(/\r$/,"")
				if (str2a[count] == ""){
					str2a[count] = $0
				}else{
					str2a[count] = sprintf("%s\n%s", str2a[count], $0)
				}
			}
			ret = getline <ENVIRON["CGICTRL_PARAM"]
		}
	}
	{
		for (i = 0; i <= count; i++){
			gsub(str1a1[i],str2a[i])
			gsub(str1a2[i],str2a[i])
		}
		s = sprintf("%s[^=%s]*%s",start,end,end)
		gsub(s,"")
		s = sprintf("%s[^=%s]*=",start,end)
		gsub(s,"")
		gsub(end,"")
		print
	}
	END{
	}'

	rm "${CGICTRL_PARAM}"
}
#
# Usage: getrealres resourceid
# Get real resouce in string
#
function getrealres {
	res="$1"
	realres=`cat "${CGICTRL_HOMEDIR}/resource/${res}" | egrep -v '^#' | sed -e '/^$/d'`
	echo "${realres}"
}
#
# Usage: getspa spaname
# Get value of spaname
#
function getspa {
	spaname="$1"
	value=`cat "${CGICTRL_TMPDIR}/spa/${CGICTRL_SESSIONID}/${spaname}" 2>/dev/null`
	echo "${value}"
}
#
# Usage: setspa spaname value
# Set spaname to value
#
function setspa {
	spaname="$1"
	value="$2"
	echo "${value}" >"${CGICTRL_TMPDIR}/spa/${CGICTRL_SESSIONID}/${spaname}"
}
#
# Usage: clearallspa
# Clear all spa in current session
#
function clearallspa {
	rm -f "${CGICTRL_TMPDIR}/spa/${CGICTRL_SESSIONID}"/*
}
#
# Usage: clearspa spaname
# Clear spaname in current session
#
function clearspa {
	spaname="$1"
	rm -f "${CGICTRL_TMPDIR}/spa/${CGICTRL_SESSIONID}/${spaname}"
}
#
# Usage: isoutputmode
# Get current output mode
#
function isoutputmode {
	if [ "X${CGICTRL_IOMODE}" = "Xoutput" ]
	then
		echo "${CGICTRL_IOMODE}"
		return 0
	else
		return 1
	fi
}
#
# Usage: isinputmode
# Get current input mode
#
function isinputmode {
	if [ "X${CGICTRL_IOMODE}" = "Xinput" ]
	then
		echo "${CGICTRL_IOMODE}"
		return 0
	else
		return 1
	fi
}
#
# Usage: getuser
# Get authenticated user name
#
function getuser {
	if [ -z "${REMOTE_USER}" ]
	then
		echo "${CGICTRL_NOAUTHENTICATEDUSER}"
	else
		echo "${REMOTE_USER}"
	fi
}
#
# Usage: settran transaction_name
# Set CGICTRL_TRAN in spa to transaction_name
#
function settran {
	tran="$1"
	setspa CGICTRL_TRAN "${tran}"
}
#
# Usage: gettran
# Get current transaction name
#
function gettran {
	getspa CGICTRL_TRAN
}
#
# Usage: ispermitted
# Check current user and current transaction are permitted
#
function ispermitted {
	user=`getuser`
	tran=`gettran`
	ispermittedwith "${user}" "${tran}"
	return $?
}
#
# Usage: ispermittedwith USER TRANSACTION
# Check specified user and ransaction are permitted
#
function ispermittedwith {
	user="$1"
	tran="$2"
	for i in `cat "${CGICTRL_HOMEDIR}/usertran/${user}" 2>/dev/null | egrep -v '^#' | sed -e '/^$/d'`
	do
		if [ `echo "${tran}" | egrep "$i"` ]
		then
			echo "${user}:${tran}"
			return 0
		fi
	done
	return 1
}
